homeassistant/components/smartthings/__init__.py

```diff
@@ -2,416 +2,140 @@
 
-CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
 
+@dataclass
+class SmartThingsData:
+    """Define an object to hold SmartThings data."""
 
-async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Initialize the SmartThings platform."""
-    await setup_smartapp_endpoint(hass, False)
-    return True
+    devices: dict[str, FullDevice]
+    scenes: dict[str, Scene]
+    client: SmartThings
 
 
-async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-    """Handle migration of a previous version config entry.
-
-    A config entry created under a previous version must go through the
-    integration setup again so we can properly retrieve the needed data
-    elements. Force this by removing the entry and triggering a new flow.
-    """
-    # Remove the entry which will invoke the callback to delete the app.
-    hass.async_create_task(hass.config_entries.async_remove(entry.entry_id))
-    # only create new flow if there isn't a pending one for SmartThings.
-    if not hass.config_entries.flow.async_progress_by_handler(DOMAIN):
-        hass.async_create_task(
-            hass.config_entries.flow.async_init(
-                DOMAIN, context={"source": SOURCE_IMPORT}
-            )
-        )
-
-    # Return False because it could not be migrated.
-    return False
-
-
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+@dataclass
+class FullDevice:
+    """Define an object to hold device data."""
+
+    device: Device
+    status: dict[str, dict[Capability, dict[Attribute, Status]]]
+
+
+type SmartThingsConfigEntry = ConfigEntry[SmartThingsData]
+
+PLATFORMS = [
+    Platform.BINARY_SENSOR,
+    Platform.CLIMATE,
+    Platform.COVER,
+    Platform.FAN,
+    Platform.LIGHT,
+    Platform.LOCK,
+    Platform.SCENE,
+    Platform.SENSOR,
+    Platform.SWITCH,
+]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: SmartThingsConfigEntry) -> bool:
     """Initialize config entry which represents an installed SmartApp."""
-    # For backwards compat
-    if entry.unique_id is None:
-        hass.config_entries.async_update_entry(
-            entry,
-            unique_id=format_unique_id(
-                entry.data[CONF_APP_ID], entry.data[CONF_LOCATION_ID]
-            ),
-        )
-
-    if not validate_webhook_requirements(hass):
-        _LOGGER.warning(
-            "The 'base_url' of the 'http' integration must be configured and start with"
-            " 'https://'"
-        )
-        return False
-
-    api = SmartThings(async_get_clientsession(hass), entry.data[CONF_ACCESS_TOKEN])
-
-    # Ensure platform modules are loaded since the DeviceBroker will
-    # import them below and we want them to be cached ahead of time
-    # so the integration does not do blocking I/O in the event loop
-    # to import the modules.
-    await async_get_loaded_integration(hass, DOMAIN).async_get_platforms(PLATFORMS)
+    if CONF_TOKEN not in entry.data:
```

Comment:
Why do we have this check here?

Reply:
Because I can't start the reauth flow in version 2, and have the reauth fow bump it to version 3. Since that would mean the entry is in MIGRATION_ERROR state and then I can't reload the entry anymore and that requires a restart.

So this is my attempt to find out that we need to start to reauth